File: src/main/java/net/vpg/apex/core/AudioData.java
package net.vpg.apex.core;

import net.vpg.apex.Apex;
import net.vpg.apex.Util;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;
import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.Arrays;

public class AudioData {
    private final int frameLength;
    private final AudioFormat format;
    private final byte[] data;
    private AudioInputStream stream;
    private int readPos;
    private int cachedPos;
    private boolean caching;

    public AudioData(File file, int frameLength) throws UnsupportedAudioFileException, IOException {
        this(AudioSystem.getAudioInputStream(file), frameLength);
    }

    public AudioData(URL url, int frameLength) throws UnsupportedAudioFileException, IOException {
        this(AudioSystem.getAudioInputStream(url), frameLength);
    }

    public AudioData(AudioInputStream sourceStream, int frameLength) {
        AudioFormat sourceFormat = sourceStream.getFormat();
        format = new AudioFormat(
            AudioFormat.Encoding.PCM_SIGNED,
            sourceFormat.getSampleRate(),
            16,
            sourceFormat.getChannels(),
            4,
            sourceFormat.getSampleRate(), // Note: Keep Sample Rate = Frame Rate
            sourceFormat.isBigEndian()
        );
        stream = AudioSystem.getAudioInputStream(format, sourceStream);
        data = new byte[frameLength * format.getFrameSize()];
        this.frameLength = frameLength;
    }

    public int getFrameLength() {
        return frameLength;
    }

    public AudioFormat getFormat() {
        return format;
    }

    public int getReadPos() {
        return readPos;
    }

    public void setReadPos(int readPos) {
        this.readPos = readPos;
    }

    public void startCaching() {
        if (caching || stream == null)
            return;
        caching = true;
        Apex.EXECUTOR.execute(() -> Util.run(this::cache));
    }

    public void stopCaching() {
        caching = false;
    }

    public byte[] readData(int frames) {
        int frameSize = format.getFrameSize();
        if (readPos + frames < cachedPos) {
            readPos += frames;
            return Arrays.copyOfRange(data, readPos * frameSize, (readPos + frames) * frameSize);
        } else if (caching) {
            Util.sleep(25);
            return readData(frames);
        }
        System.out.println("Error!");
        throw new IllegalStateException();
    }

    public void cache() throws IOException {
        int frameSize = stream.getFormat().getFrameSize();
        int off = 0, read;
        while (caching && (read = stream.read(data, off, data.length)) != -1) {
            off += read;
            cachedPos = off / frameSize;
        }
        caching = false;
        stream.close();
        stream = null;
    }
}

File: src/main/java/net/vpg/apex/core/AudioData.kt
/*
 * Copyright 2021 Vaibhav Nargwani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.vpg.apex.core

import net.vpg.apex.Apex
import java.io.File
import java.net.URL
import javax.sound.sampled.AudioInputStream
import javax.sound.sampled.AudioSystem

class AudioData(private val stream: AudioInputStream, val frameLength: Int) {
    private val data = ByteArray(frameLength * format.frameSize)
    var readPos = 0
    private var cachedPos = 0
    private var caching = false

    constructor(file: File, frameLength: Int) : this(AudioSystem.getAudioInputStream(file), frameLength)

    constructor(url: URL, frameLength: Int) : this(AudioSystem.getAudioInputStream(url), frameLength)

    val format
        get() = stream.format

    fun startCaching() {
        if (caching) return
        caching = true
        Apex.EXECUTOR.execute { cache() }
    }

    fun readData(frames: Int): ByteArray {
        if (readPos + frames <= cachedPos) {
            readPos += frames
            return data.copyOfRange(readPos * format.frameSize, (readPos + frames) * format.frameSize)
        }
        if (caching) {
            Thread.sleep(25)
            return readData(frames)
        }
        throw IllegalStateException()
    }

    fun cache() {
        var off = 0
        while (caching && (stream.read(data, off, data.size - off).also { off += it }) != -1) {
            cachedPos = off / format.frameSize
        }
        close()
    }

    fun close() {
        caching = false
        stream.close()
    }
}
